File: Ser2023/Store/Store.Web/Modules/ServerTypes/Store/WaresForm.ts
import { CategoriesDialog } from "@/Store/Categories/CategoriesDialog";
import { LookupEditor, StringEditor, IntegerEditor, ImageUploadEditor, BooleanEditor, DecimalEditor, PrefixedContext } from "@serenity-is/corelib";
import { initFormType } from "@serenity-is/corelib/q";

export interface WaresForm {
    CategoryId: LookupEditor;
    WaresCode: StringEditor;
    WaresBarcode: StringEditor;
    WaresLabel: StringEditor;
    WaresName: StringEditor;
    MeasureId: IntegerEditor;
    WaresImage: ImageUploadEditor;
    Discontinued: BooleanEditor;
    CounterpartyId: StringEditor;
    QuantityPerUnit: DecimalEditor;
    UnitPrice: DecimalEditor;
    AccountId: IntegerEditor;
}

export class WaresForm extends PrefixedContext {
    static formKey = 'Store.Wares';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!WaresForm.init)  {
            WaresForm.init = true;

            var w0 = LookupEditor;
            var w1 = StringEditor;
            var w2 = IntegerEditor;
            var w3 = ImageUploadEditor;
            var w4 = BooleanEditor;
            var w5 = DecimalEditor;

            initFormType(WaresForm, [
                'CategoryId', w0,
                'WaresCode', w1,
                'WaresBarcode', w1,
                'WaresLabel', w1,
                'WaresName', w1,
                'MeasureId', w2,
                'WaresImage', w3,
                'Discontinued', w4,
                'CounterpartyId', w1,
                'QuantityPerUnit', w5,
                'UnitPrice', w5,
                'AccountId', w2
            ]);
        }
    }
}

[CategoriesDialog]; // referenced types

File: Ser2023/Store/Store.Web/Modules/Store/Wares/WaresForm.cs
using Serenity.ComponentModel;
using Serenity.Web;
using System;
using System.ComponentModel;

namespace Store.Store.Forms
{
    [FormScript("Store.Wares")]
    [BasedOnRow(typeof(Entities.WaresRow), CheckNames = true)]
    public class WaresForm
    {
        [Category("General")]
        public Int32 CategoryId { get; set; }
        public String WaresCode { get; set; }
        public String WaresBarcode { get; set; }
        public String WaresLabel { get; set; }
        public String WaresName { get; set; }
        public Int32 MeasureId { get; set; }
        public String WaresImage { get; set; }
        public Boolean Discontinued { get; set; }
        public String CounterpartyId { get; set; }

        [Category("Pricing")]
        public Decimal QuantityPerUnit { get; set; }
        public Decimal UnitPrice { get; set; }

        public Int32 AccountId { get; set; }
    }
}